sql: wait to start index GC job until schema changer is done

thoszhang opened the pull request:
Previously, we were creating and starting the index GC job in a separate
transaction from the one used in `PublishMultiple()` in
`SchemaChanger.done()`, meaning that the index GC job could be run
before the original transaction to finalize all schema changes on the
table descriptor had been committed. This led to unpredictable behavior
while testing. It could also potentially cause multiple GC jobs to be
created if the `PublishMultiple()` closure were retried.

In this PR, the GC job is now created as a `StartableJob` that is
started only after the table descriptor in the finalized state has been
published.

Release justification: Bug fix for new feature.

Release note (bug fix): Fixed a bug introduced in beta.3 that could
cause multiple index GC jobs to be created for the same schema change in
rare cases.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/46929)
<!-- Reviewable:end -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46929)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner, @lucy-zhang, and @pbardea)

___
*[pkg/sql/schema_changer.go, line 981 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46929#-M3vl715BC8qfyU3J20n:-M3vl715BC8qfyU3J20o:b-2fhsy3) ([raw file](https://github.com/cockroachdb/cockroach/blob/99a26157a3e4362c66250de1e69ca32bcb6a1e15/pkg/sql/schema_changer.go#L981)):*
> ```Go
> 		)
> 	})
> 	if err != nil {
> ```

can you add:
```
if indexGCJob != nil {
    if rollbackErr := indexGCJob.CleanupOnRollback(ctx); rollbackErr != nil {
        log.Warningf(ctx, "failed to cleanup job: %v", err)
    }
}
```
___
*[pkg/sql/schema_changer.go, line 984 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46929#-M3vkE2-CprjfzIjbvd9:-M3vkE203w4NVovywREt:b-oxe5tr) ([raw file](https://github.com/cockroachdb/cockroach/blob/99a26157a3e4362c66250de1e69ca32bcb6a1e15/pkg/sql/schema_changer.go#L984)):*
> ```Go
> 		return nil, err
> 	}
> 	// TODO (lucy): Can we do the same thing with the PK drop index job?
> ```

same thing  meaning... I get that this may make sense in the context of this PR but in general this TODO isn't adequately detailed. 


<!-- Sent from Reviewable.io -->

thoszhang reviewed (commented):
I screwed up and treated the index GC job as a single job, when in fact there can be multiple indexes per transaction to be dropped. Now it's a slice of jobs.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46929#-:-M3wF1DW92lTeXSkhYfp:bplxj3w)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner and @pbardea)

___
*[pkg/sql/schema_changer.go, line 981 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46929#-M3vl715BC8qfyU3J20n:-M3wEGFA-wwTPLLUkMoE:bfeb0v0) ([raw file](https://github.com/cockroachdb/cockroach/blob/99a26157a3e4362c66250de1e69ca32bcb6a1e15/pkg/sql/schema_changer.go#L981)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

can you add:
```
if indexGCJob != nil {
    if rollbackErr := indexGCJob.CleanupOnRollback(ctx); rollbackErr != nil {
        log.Warningf(ctx, "failed to cleanup job: %v", err)
    }
}
```
</blockquote></details>

Done. Unfortunately we still won't run the cleanup if `Publish()` has to retry, but I guess this is a best-effort thing anyway.
___
*[pkg/sql/schema_changer.go, line 984 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46929#-M3vkE2-CprjfzIjbvd9:-M3wEGrn9FkaSg_de9qK:berc48p) ([raw file](https://github.com/cockroachdb/cockroach/blob/99a26157a3e4362c66250de1e69ca32bcb6a1e15/pkg/sql/schema_changer.go#L984)):*
<details><summary><i>Previously, ajwerner wrote…</i></summary><blockquote>

same thing  meaning... I get that this may make sense in the context of this PR but in general this TODO isn't adequately detailed. 
</blockquote></details>

Done. (I moved it to where we actually create that index drop job.)


<!-- Sent from Reviewable.io -->

ajwerner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46929#-:-M3wnjthDPNeRlfz9Yh1:bnfp4nl)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @pbardea)

___
*[pkg/sql/schema_changer.go, line 981 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/46929#-M3vl715BC8qfyU3J20n:-M3wnO0P5oJ9CPcRO802:b-b6y7ww) ([raw file](https://github.com/cockroachdb/cockroach/blob/99a26157a3e4362c66250de1e69ca32bcb6a1e15/pkg/sql/schema_changer.go#L981)):*
<details><summary><i>Previously, lucy-zhang (Lucy Zhang) wrote…</i></summary><blockquote>

Done. Unfortunately we still won't run the cleanup if `Publish()` has to retry, but I guess this is a best-effort thing anyway.
</blockquote></details>

yeah... about that.. I've had that on my TODO list for a bit now. 

The non-idempotent part of all of this is the fact that we create a new job id every time. If we created the ID above the call to create then it'd all be good. I'll go refactor this eventually. Until then it's potentially a very slow leak that does make me sad.

Or in this case I think that would look like having a slice of ids which we allocate the first pass through and then reuse on retries. That's a bit nasty but I don't have a better answer.


<!-- Sent from Reviewable.io -->

pbardea reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> 
I left a comment, but if you think it would be best to fix that in a separate PR I'm happy with that too.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46929#-:-M3wpdpmAtAqqQYPayj9:b-pqy6nx)* status: :shipit: complete! 2 of 0 LGTMs obtained (waiting on @lucy-zhang)

___
*[pkg/sql/schema_changer.go, line 1415 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46929#-M3wot7x0iJ05GsAN1UH:-M3wot7x0iJ05GsAN1UI:b-wuf8mb) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc274b05272240614d79bd2bd47a84d29aa5d26/pkg/sql/schema_changer.go#L1415)):*
> ```Go
> 			descriptorIDs = append(descriptorIDs, table.ID)
> 		}
> 	}
> ```

I think we may have a missing case here to properly handle the empty database case. Something like
```
} else if details.ParentID != sqlbase.InvalidID {
  descriptorIDs = []sqlbase.ID{details.ParentID}
}
```
so that jobs dropping empty databases aren't cleaned up accidentally.

Also happy to dig into the drop empty database case in a separate PR.


<!-- Sent from Reviewable.io -->

lucy-zhang committed:
sql: wait to start index GC job until schema changer is done

Previously, we were creating and starting the index GC job in a separate
transaction from the one used in `PublishMultiple()` in
`SchemaChanger.done()`, meaning that the index GC job could be run
before the original transaction to finalize all schema changes on the
table descriptor had been committed. This led to unpredictable behavior
while testing. It could also potentially cause multiple GC jobs to be
created if the `PublishMultiple()` closure were retried.

In this PR, the GC job is now created as a `StartableJob` that is
started only after the table descriptor in the finalized state has been
published.

Release note (bug fix): Fixed a bug introduced in 20.1 that could cause
multiple index GC jobs to be created for the same schema change in rare
cases.

thoszhang reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/46929)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 2 stale) (waiting on @ajwerner and @pbardea)

___
*[pkg/sql/schema_changer.go, line 1415 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/46929#-M3wot7x0iJ05GsAN1UH:-M4vgacm-tx0i8be1REW:b-ubtdtn) ([raw file](https://github.com/cockroachdb/cockroach/blob/5bc274b05272240614d79bd2bd47a84d29aa5d26/pkg/sql/schema_changer.go#L1415)):*
<details><summary><i>Previously, pbardea (Paul Bardea) wrote…</i></summary><blockquote>

I think we may have a missing case here to properly handle the empty database case. Something like
```
} else if details.ParentID != sqlbase.InvalidID {
  descriptorIDs = []sqlbase.ID{details.ParentID}
}
```
so that jobs dropping empty databases aren't cleaned up accidentally.

Also happy to dig into the drop empty database case in a separate PR.
</blockquote></details>

Dropping an empty database doesn't cause a GC job to be created (either in 19.2 or now; in 19.2, there's no job at all).

I think an earlier iteration of this PR might have accidentally changed this (or maybe that was on my local branch, I don't remember). But in any case I've now added a test to ensure there's no GC job for an empty database.


<!-- Sent from Reviewable.io -->

lucy-zhang commented:
Hmm. I managed to also repro the `TestProtectedTimestampsDuringBackup` CI flake on master, on my laptop, but the results were a bit weird: in one attempt it took >20k runs over 6 hours (I left it running overnight) and in another attempt my computer went to sleep in the middle, which might have affected the timeouts or something. I'm going to try again on roachprod-stress and open an issue once I see what happens in a more controlled environment. (This failure is different from #45932.)

lucy-zhang commented:
Filed #47522 for the `TestProtectedTimestampsDuringBackup` flake (and #47532, which I encountered in testing), merging now.

bors r+

lucy-zhang commented:
The flake was https://github.com/cockroachdb/cockroach/issues/47546.

bors r+

lucy-zhang commented:
This `TestScrubFKConstraintFKNulls` flake is persistent. I'll revisit this later.

lucy-zhang commented:
bors r+

lucy-zhang commented:
???

bors r+
